modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -290,37 +297,58 @@ public CompletableFuture<Void> finish(
             return completedFuture(null);
         }
 
-        observableTimestampTracker.update(commitTimestamp);
-
         Function<Void, CompletableFuture<Void>> clo = ignored -> {
-            TxFinishReplicaRequest req = FACTORY.txFinishReplicaRequest()
-                    .txId(txId)
-                    .timestampLong(clock.nowLong())
-                    .groupId(commitPartition)
-                    .groups(groups)
-                    .commit(commit)
-                    .commitTimestampLong(hybridTimestampToLong(commitTimestamp))
-                    .term(term)
-                    .build();
-
-            return replicaService.invoke(recipientNode, req)
-                    .thenRun(() ->
-                            updateTxMeta(txId, old -> {
-                                if (isFinalState(old.txState())) {
-                                    finishingStateMeta.txFinishFuture().complete(old);
-
-                                    return old;
-                                }
-
-                                assert old instanceof TxStateMetaFinishing;
-
-                                TxStateMeta finalTxStateMeta = coordinatorFinalTxStateMeta(commit, commitTimestamp);
-
-                                finishingStateMeta.txFinishFuture().complete(finalTxStateMeta);
-
-                                return finalTxStateMeta;
-                            })
-                    );
+            // In case of commit it's required to check whether current primaries are still the same that were enlisted and whether
+            // given primaries are not expired or, in other words, whether commitTimestamp is less or equal to the enlisted primaries
+            // expiration timestamps.
+            CompletableFuture<Void> verificationFuture =
+                    commit ? verifyCommitTimestamp(enlistedGroups, commitTimestamp) : completedFuture(null);
+
+            return verificationFuture.handle(
+                    (unused, throwable) -> {
+                        // TODO: https://issues.apache.org/jira/browse/IGNITE-19170 Use ZonePartitionIdMessage and remove cast
+                        Collection<ReplicationGroupId> replicationGroupIds = new HashSet<>();
+                        replicationGroupIds.addAll((Collection<ReplicationGroupId>) (Collection<?>) enlistedGroups.keySet());
+
+                        boolean verifiedCommit = throwable == null && commit;
```

Comment:
I do not understand where we will be able to see this exception, which changed the transaction state (to rollback) at the last moment.

Reply:
Please check
```
                    // verification future is added in order to share proper exception with the client
                    .thenCompose(r -> verificationFuture);
```
at the end of the future chaining.

Reply:
Tricky way, I cannot insist, but I would make this code more clear.

Reply:
Well, it's a common pattern of "re-throwing" an exception from a future after handling it.

Reply:
I would create another future that is used to return.

Reply:
Again, proposed solution is a common pattern.